wallet/src/key.rs

```diff
@@ -78,22 +80,130 @@ where
+        Ok(ExtendedSK {
+            secret_key,
+            chain_code,
+        })
     }
 }
 
+#[derive(Debug, PartialEq, Fail)]
+pub enum KeyDerivationError {
+    /// Invalid hmac key length
+    #[fail(display = "The length of the hmac key is invalid")]
+    InvalidKeyLength,
+    /// Invalid seed length
+    #[fail(display = "The length of the seed is invalid, must be between 128/256 bits")]
+    InvalidSeedLength,
+
+    #[fail(display = "Error in secp256k1 crate")]
+    Secp256k1Error,
+}
+
 /// Extended Key is just a Key with a Chain Code
-pub struct ExtendedKey<K> {
-    pub secret: K,
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct ExtendedSK {
+    pub secret_key: SecretKey,
     pub chain_code: [u8; 32],
 }
 
+/// A child number for a derived key
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct ChildNumber(u32);
+
+impl ChildNumber {
+    pub fn is_hardened(&self) -> bool {
+        self.0 & HARDENED_BIT == HARDENED_BIT
+    }
+
+    pub fn is_normal(&self) -> bool {
+        self.0 & HARDENED_BIT == 0
+    }
+
+    pub fn to_bytes(&self) -> [u8; 4] {
+        self.0.to_be_bytes()
+    }
+}
+
+impl ExtendedSK {
+    /// Try to derive an extended private key from a given path
+    pub fn derive(seed: &[u8], path: Vec<ChildNumber>) -> Result<ExtendedSK, KeyDerivationError> {
+        let key_bytes = DEFAULT_HMAC_KEY.as_ref();
+        let mut hmac512 = Hmac::<sha2::Sha512>::new_varkey(key_bytes)
+            .map_err(|_| KeyDerivationError::InvalidKeyLength)?;
+        hmac512.input(seed);
+
+        let i = hmac512.result().code();
+        let (il, ir) = i.split_at(32);
+        let chain_code: [u8; 32] = {
+            let mut array: [u8; 32] = [0; 32];
+            array.copy_from_slice(&ir);
+            array
+        };
+        let secret_key =
+            SecretKey::from_slice(&il).map_err(|_| KeyDerivationError::Secp256k1Error)?;
+
+        let mut extended_sk = ExtendedSK {
+            secret_key,
+            chain_code,
+        };
+
+        for child in path {
+            extended_sk = extended_sk.child(child)?
+        }
+
+        Ok(extended_sk)
+    }
+
+    pub fn secret(&self) -> [u8; 32] {
+        let mut secret: [u8; 32] = [0; 32];
+        secret.copy_from_slice(&self.secret_key[..]);
+
+        secret
+    }
+
+    /// Try to get a private child key from parent
+    pub fn child(&self, child: ChildNumber) -> Result<ExtendedSK, KeyDerivationError> {
+        let mut hmac512: Hmac<sha2::Sha512> =
+            Hmac::new_varkey(&self.chain_code).map_err(|_| KeyDerivationError::InvalidKeyLength)?;
+        if child.is_hardened() {
+            hmac512.input(&[0]);
+            hmac512.input(&self.secret_key[..]);
+        } else {
+            hmac512.input(
+                &PublicKey::from_secret_key(&secp256k1::Secp256k1::new(), &self.secret_key)
+                    .serialize(),
+            );
+        }
+        hmac512.input(&child.to_bytes());
+        let i = hmac512.result().code();
```

Comment:
This chunk of code is repeated in `derive` and `child`, so it should be extracted into a helper function.